README.txt
==============================
Belief Propagation in Networks
==============================

Belief propagation is a general graph algorithm for ranking nodes
based on their location in the graph and some prior knowledge. 

The network is using in machine learning and bioinformatics in a variety 
of applications.

This package implements the algorithm and provides a command line interface 
to run it. 

Dependencies:

* numpy 
* scipy




CHANGES.txt
v0.2.1, July 3rd 2014  -- * The quality of the results can now be evaluated with a validation set (requires scikit.learn)

v0.2.0, June 25th 2014 -- * The code in now split into modules and can be used as a library.
                          * Support for creation of data structures from pandas dataframes 
						  * Support for cross validation for the prediction of a given prior

v0.1.0, June 16th 2014 -- Initial release


propagate/cv.py
import numpy as np 
from sklearn.metrics import roc_curve, auc

import command_line
import data

def parse_commandline(args):
    parser = command_line.create_parser()

    parser.add_argument('-s,--seed', dest="seed", type=int, default=None, help="a seed for the random number generator (default: use internal logic)")
    parser.add_argument('-f,--folds', dest="folds", type=int, default=5, help="the number of cross-validation folds")
    parser.add_argument('-l,--leave-one-out', dest="loo", action='store_true', default=False, help="Perform leave-one-out cross validation")

    return parser.parse_args(args)

def leave_one_out(network, prior, **kwargs):
    prior_ind = np.nonzero(prior)[0]
    result = np.zeros(shape=(len(prior_ind),len(prior)))

    prior_cpy = np.copy(prior)

    for i in range(len(prior_ind)):
        prior_cpy[prior_ind[i]] = 0
        if i > 0:
            prior_cpy[prior_ind[i-1]] = prior[prior_ind[i-1]]

        result[i,:] = network.smooth(prior_cpy, **kwargs).flat
        result[i,np.where((prior_cpy>0).flat)] = None

    return result

def kfold(network, prior, folds, seed=None, **kwargs):
    # shuffle the non zero indices so that each cv iteration 
    # is contiguous in prior_ind
    np.random.seed(seed)
    prior_ind = np.copy(np.nonzero(prior)[0])
    np.random.shuffle(prior_ind)
    
    if folds > len(prior_ind):
        raise ValueError("Folds should be less than the number of elements in the prior")
    
    result = np.zeros(shape=(folds, len(prior)))
    cv_chunk = len(prior_ind) // folds

    for i in range(folds):
        prior_cpy = np.copy(prior)
        if i < folds-1:
            prior_cpy[prior_ind[(i*cv_chunk):((i+1)*cv_chunk)]] = 0 # remove associations
        else:
            prior_cpy[prior_ind[(i*cv_chunk):]] = 0 # remove associations
        result[i,:] = network.smooth(prior_cpy, **kwargs).flat
        result[i,np.where((prior_cpy>0).flat)] = None

    return result

def roc(result, prior, aggregate=np.nanmean):
    scores = aggregate(result, axis=0)
    fpr, tpr, _ = roc_curve(prior>0, scores) 
    return tpr, fpr, auc(fpr, tpr)

if __name__ == '__main__':
    import sys
    
    opts = parse_commandline(sys.argv[1:])
    succeeded = command_line.is_valid_args(opts)
    
    if opts.prior is None:
        print "Prior filename must be specfied for cross validation"
        succeeded = False

    if opts.folds <= 1:
        print "Folds should be greater than 2"
        succeeded = False

    if succeeded:
        network = data.read_ppi(opts.network)
        prior = data.read_prior(opts.prior, network.names)

        if np.any(prior > 1.0):
            sys.stderr.write("Scaling prior, values outside [0,1]\n")
            prior /= np.max(prior)

        network = network.normalize()

        if opts.loo:
            result = leave_one_out(network, prior, alpha=opts.alpha, eps=opts.epsilon, max_iter=opts.iterations)
        else:
            result = kfold(network, prior, opts.folds, opts.seed, alpha=opts.alpha, eps=opts.epsilon, max_iter=opts.iterations)

        # handle the results array 
        tpr, fpr, auc_score = roc(result, prior)

        print tpr
        print fpr
        print auc_score


propagate/run.py
import os.path

import numpy as np

import data
import command_line

def parse_commandline(args):
    parser = command_line.create_parser()

    parser.add_argument('--validation', dest="validation", type=str, default=None, help="A validation dataset, that is use 'prior' +'network' to predict 'validation'")
    return parser.parse_args(args)

def validate(result, validation):
    import sklearn.metrics as metrics
    fpr, tpr, _ = metrics.roc_curve(validation>0, result)
    return tpr, fpr, metrics.auc(fpr, tpr)

if __name__ == '__main__':
    import sys
    
    opts = parse_commandline(sys.argv[1:])
    succeeded = command_line.is_valid_args(opts)

    if opts.validation is not None and not os.path.exists(opts.validation):
        succeeded = False
        print "Invalid validation filename"
    
    if succeeded:
        network = data.read_ppi(opts.network)
        if opts.prior is not None:
            prior = data.read_prior(opts.prior, network.names)
            if np.any(prior > 1.0):
                sys.stderr.write("Scaling prior, values outside [0,1]\n")
                prior /= np.max(prior)
        else:
            prior = np.ones((network.graph.shape[0],1))

        if opts.validation is not None:
            validation = data.read_prior(opts.validation, network.names)
          
        result = network.normalize().smooth(prior,
                                            alpha=opts.alpha,
                                            eps=opts.epsilon,
                                            max_iter=opts.iterations)

        if opts.validation is not None:
            tpr, fpr, auc = validate(result, validation)
            print "fpr: ", fpr
            print "tpr: ", tpr
            print "auc: ", auc
        else:
            for item in sorted(zip(network.names.index, result.flat), key=lambda x: x[1]):
                print "\t".join(map(str, item))


propagate/data.py
import numpy as np
import pandas as pd 
import scipy.sparse as sps
from network import Network

def _build_index(dataset, *columns):
    mapset = pd.DataFrame(columns=('name', 'index'), dtype=(str,int))
    for name_col, idx_col in columns:
        subset = dataset[[name_col, idx_col]]
        subset.columns = 'name', 'index'
        mapset = mapset.append(subset, ignore_index=True)
    
    mapset.drop_duplicates(inplace=True)
    #mapset['name'] = mapset.name.astype(str)
    return mapset

def read_ppi_from_dataframe(dataset, from_column='from', to_column='to', confidence_column='confidence'):
    all_genes = pd.concat((dataset[from_column], dataset[to_column])).unique()
    n = len(all_genes)

    names = pd.DataFrame(data={'name': all_genes, 'index': range(n)})
    
    d1 = dataset.merge(names, left_on=from_column, right_on='name')
    d1.rename(columns={'index': 'from_index'}, inplace=True)
    d1.drop('name', axis=1, inplace=True)
    d2 = d1.merge(names, left_on=to_column, right_on='name')
    d2.rename(columns={'index': 'to_index'}, inplace=True)
    d2.drop('name', axis=1, inplace=True)

    d2[confidence_column].fillna(1, inplace=True)
    d = sps.coo_matrix((d2[confidence_column], (d2['from_index'], d2['to_index'])), shape=(n,n))
    sym = (d + d.T)
    
    names = _build_index(d2, (from_column, 'from_index'), (to_column, 'to_index'))
    names.sort(columns=('index'), inplace=True)

    # for some strange reason the following:
    # > names_series = pd.Series(index=names['name'].values, data=names['index'].values)
    # will convert the index to int, while the current version won't.
    # it does not reproduce in a 'simple' case.
    names_series = pd.Series(index=names.name.values, data=names['index'].values)
    return Network(sym, names_series)

def read_ppi(filename):
    """ reads a ppi network from file  """
        
    column_names = ('from', 'to', 'confidence', 'flag')
    dataset = pd.read_table(filename, header=None, names=column_names)

    return read_ppi_from_dataframe(dataset)

def create_prior(index, names, scores):
    prior = np.zeros(shape=(index.shape[0], 1))
    mapped_names = index[names]
    na_mask = (~np.isnan(mapped_names)).values
    if hasattr(scores, '__getitem__'):
        prior[mapped_names[na_mask].values.astype(np.int)] = scores[na_mask].reshape((np.count_nonzero(na_mask),1))
    else:
        prior[mapped_names[na_mask].values.astype(np.int)] = scores

    return prior

def read_prior(filename, index):
    prior = np.zeros((len(index),1))
    with open(filename, 'r') as f:
        for line in f:
            name, score = line.strip().split()
            prior[index[name]] = float(score)

    return prior




propagate/command_line.py
import argparse

def create_parser():
    parser = argparse.ArgumentParser(description="Compute a smoothed score over network nodes")

    # positional
    parser.add_argument('network', type=str, help="network filename (tsv format: from/to/weight)")
    parser.add_argument('prior', type=str, default=None, help="prior filename (tsv format: from/value)")
    # named
    parser.add_argument('-a,--alpha', dest="alpha", type=float, default=0.6, help="relative weight for the network")
    parser.add_argument('-i,--iterations', dest="iterations", type=int, default=1000, help="maximum number of iterations to execute")
    parser.add_argument('-e,--epsilon', dest="epsilon", type=float, default=1e-5, help="convergence threshold")
    parser.add_argument('-v,--verbose', dest="verbose", action='store_true', default=False, help="verbose output")

    return parser

def is_valid_args(opts):
    import sys
    import os.path
    
    if opts.verbose:
        sys.stderr.write(str(opts) + "\n")
    
    failed = False
    if opts.network is None or not os.path.exists(opts.network):
        print "Invalid network filename"
        failed = True

    if opts.prior is not None and not os.path.exists(opts.prior):
        print "Invalid prior filename"
        failed = True

    if opts.epsilon > 0.5:
        print "Epsilon is too large"
        failed = True
    
    if not 0 < opts.alpha < 1:
        print "alpha should be in (0,1)"
        failed = True

    return not failed




propagation/propagation.py
#! /usr/bin/env python

import numpy as np 
import numpy.linalg as la
import pandas as pd
import scipy.sparse as sps
import os.path

class Network: 
    def __init__(self, graph, names):
        self.graph = graph
        self.names = names

    def normalize(self):
        data = 1.0 / np.sqrt(self.graph.sum(1))
        d = sps.dia_matrix( (data.T,[0]), (len(data),len(data)) )
        self.graph = d * self.graph * d

        return self

    def smooth(self, y, alpha=0.6, eps=1e-5, max_iter=1000):
        f = np.copy(y)
        for i in range(max_iter):
            fold, f = f, alpha * self.graph * f + (1-alpha) * y
            if la.norm(f-fold) < eps:
                break

        return f

def _build_index(dataset, *columns):
    mapset = pd.DataFrame()
    for name_col, idx_col in columns:
        subset = dataset[[name_col, idx_col]]
        subset.columns = 'name', 'index'
        mapset = mapset.append(subset, ignore_index=True)
    
    mapset.drop_duplicates(inplace=True)
    return mapset

def read_ppi(filename):
    """ reads a ppi network from file  """
        
    column_names = ('from', 'to', 'confidence', 'flag')
    dataset = pd.read_table(filename, header=None, names=column_names)

    all_genes = pd.concat((dataset['from'], dataset['to'])).unique()
    n = len(all_genes)

    names = pd.DataFrame(data={'name': all_genes, 'index': range(n)})
    
    d1 = dataset.merge(names, left_on='from', right_on='name')
    d1.rename(columns={'index': 'from_index'}, inplace=True)
    d1.drop('name', axis=1, inplace=True)
    d2 = d1.merge(names, left_on='to', right_on='name')
    d2.rename(columns={'index': 'to_index'}, inplace=True)
    d2.drop('name', axis=1, inplace=True)

    d2.confidence.fillna(1, inplace=True)
    d = sps.coo_matrix((d2['confidence'], (d2['from_index'], d2['to_index'])), shape=(n,n))
    sym = (d + d.T)
    
    names = _build_index(d2, ('from', 'from_index'), ('to', 'to_index'))
    names.sort(columns=('index'), inplace=True)

    return Network(sym, pd.Series(index=names['name'].values, data=names['index'].values))

def read_prior(filename, index):
    prior = np.zeros((len(index),1))
    with open(filename, 'r') as f:
        for line in f:
            name, score = line.strip().split()
            prior[index[name]] = float(score)

    return prior


def parse_commandline(args):
    usage = """USAGE: propagation.py [options] network prior 
   -n, --network [filename]   network filename (tsv: from/to/weight)
   -p, --prior [filename]     prior (tsv: from/value)
   -i, --iterations [int]     maximum number of iterations (default: 1000)
   -e, --epsilon [double]     epsilon for convergence (default: 1e-5)
   -a, --alpha   [double]     relative weight for the network (default: 0.6)

   -h, --help                 print this help 
"""
    states = {'-n': 0, '--network': 0, 
              '-p': 1, '--prior': 1, 
              '-i': 2, '--iterations': 2, 
              '-e': 3, '--epsilon': 3, 
              '-a': 4, '--alpha': 4,
              '-v': -1, '--verbose': -1,
              '-h': -1, '--help': -1}

    opts = {'alpha': 0.6, 'epsilon': 1e-5, 'iterations': 100, 'verbose': False}
    state = 0;

    for i in args:
        if i.startswith('-'):
            if i == '-h' or i == '--help':
                print usage
                sys.exit(-1)
            if i == '-v' or i == '--verbose':
                opts['verbose'] = True
                continue
            if i not in states:
                print "Unknown option: ", i
                print usage
                sys.exit(-2)

            state = states[i]
            continue
        
        if state == 0:
            opts['network'] = i
            state = 1
            continue

        if state == 1:
            opts['prior'] = i
            if 'network' in opts:
                break
            state = 0
            continue

        if state == 2:
            opts['iterations'] = float(i)
        elif state == 3:
            opts['epsilon'] = float(i)
        elif state == 4:
            opts['alpha'] = float(i)

        if 'network' not in opts:
            state = 0 
        elif 'prior' not in opts:
            state = 1
        else:
            state = -1

    return opts

if __name__ == '__main__':
    import sys
    
    opts = parse_commandline(sys.argv[1:])
    if opts['verbose']:
        sys.stderr.write(str(opts) + "\n")
    
    fail = False
    if 'network' not in opts or not os.path.exists(opts['network']):
        print "Invalid network filename"
        fail = True

    if 'prior' in opts and not os.path.exists(opts['prior']):
        print "Invalid prior filename"
        fail = True

    if opts['epsilon'] > 0.5:
        print "Epsilon is too large"
        fail = True
    
    if not 0 < opts['alpha'] < 1:
        print "alpha should be in (0,1)"
        fail = True

    if not fail:
        network = read_ppi(opts['network'])
        if 'prior' in opts:
            prior = read_prior(opts['prior'], network.names)
            if np.any(prior > 1.0):
                sys.stderr.write("Scaling prior, values outside [0,1]\n")
                prior /= np.max(prior)
        else:
            prior = np.ones((network.graph.shape[0],1))
          
        result = network.normalize().smooth(prior,
                                            alpha=opts['alpha'],
                                            eps=opts['epsilon'],
                                            max_iter=opts['iterations'])

        for item in sorted(zip(network.names.index, result.flat), key=lambda x: x[1]):
            print "\t".join(map(str, item))
